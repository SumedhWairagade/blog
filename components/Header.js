import React from 'react';
import PropTypes from 'prop-types';
import format from 'date-fns/format';
import formatDistance from 'date-fns/formatDistance';
import findPost from '../lib/findPost';
import Button from './Button';
import webp from '../lib/webp';

const PostDate = ({ date }) => {
  const distance = formatDistance(date, new Date(), { addSuffix: true });
  const dateString = format(date, 'MMMM D, YYYY');
  return (
    <h2 title={dateString}>
      Posted {distance}
      <style jsx>{`
        h2 {
          font-size: 3rem;
        }
      `}</style>
    </h2>
  );
};

PostDate.propTypes = {
  date: PropTypes.number.isRequired,
};

const Header = props => {
  const post = findPost(props.id);
  const { title, date, image: { imageUrl } } = post;
  return (
    <header>
      <div>
        <h1>{props.title || title}</h1>
        {!props.link && date && <PostDate date={date} />}
        {props.link && <Button text="Read More" link={props.id} />}
      </div>
      <style jsx>{`
        header {
          height: 50vh;
          min-height: 500px;
          max-height: 800px;
          background: ${imageUrl
              ? `url(/static/images/posts/${webp(imageUrl).url})`
              : `url(/static/images/posts/brevite-434280.webp)`},
            ${imageUrl
              ? `url(/static/images/posts/${imageUrl})`
              : 'url(/static/images/posts/brevite-434280.jpg)'};
          background-size: cover;
          background-position: center;

          display: flex;
          justify-content: center;
          align-items: center;
          color: white;
          position: relative;
        }

        header::before {
          content: '';
          position: absolute;
          top: 0;
          left: 0;
          width: 100%;
          height: 100%;
          background: rgba(0, 0, 0, 0.6);
        }

        div {
          z-index: 1;
          text-align: center;
          max-width: 80vw;
        }

        h1 {
          font-size: 5rem;
          margin-bottom: 2rem;
        }
      `}</style>
    </header>
  );
};

Header.defaultProps = {
  link: null,
  id: null,
  title: null,
};

Header.propTypes = {
  title: PropTypes.string,
  link: PropTypes.string,
  id: PropTypes.string,
};

export default Header;
