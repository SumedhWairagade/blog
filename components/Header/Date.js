import React from 'react';
import PropTypes from 'prop-types';
import { FormattedRelative } from 'react-intl';
import { expanded } from '../../utils/Dates';

const Date = ({ date }) => (
  <h2 title={expanded(date)}>
    Posted <FormattedRelative value={date} />
    <style jsx>{`
      h2 {
        font-size: 3rem;
      }
    `}</style>
  </h2>
);

Date.propTypes = {
  date: PropTypes.number.isRequired,
};

export default Date;
