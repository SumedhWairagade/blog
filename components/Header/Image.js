import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import webp from '../../utils/webp';
import { unsplashParams } from '../../theme';

const Picture = styled.picture`
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;

  &::before {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: rgba(0, 0, 0, 0.6);
  }
`;

const Image = styled.img`
  height: 100%;
  width: 100%;
  object-fit: cover;
`;

const HeaderImage = ({ image: { imageUrl, name, url } }) => {
  const image = `/static/images/posts/${imageUrl}`;

  const webpbg = `/static/images/posts/${webp(imageUrl).url}`;
  const mimetype = `image/${webp(imageUrl).type}`;

  return (
    <Picture>
      <source srcSet={webpbg} type="image/webp" />
      <source srcSet={image} type={mimetype} />
      <Image
        src={`/static/images/posts/${imageUrl}`}
        alt={name ? `Taken by ${name}` : ''}
        data-source-url={unsplashParams(url)}
      />
    </Picture>
  );
};

HeaderImage.propTypes = {
  image: PropTypes.shape({
    imageUrl: PropTypes.string,
    name: PropTypes.string,
    url: PropTypes.string,
  }).isRequired,
};

export default HeaderImage;
