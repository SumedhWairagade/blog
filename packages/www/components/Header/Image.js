// @flow
import React from 'react';
import styled from 'styled-components';
import { cloudinary } from '../../config';
import unsplash from '../../utils/unsplash';
import type { ImageType } from '../../types/Post';

const ImageWrap = styled.div`
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;

  &::before {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: rgba(0, 0, 0, 0.6);
  }

  img {
    height: 100%;
    width: 100%;
    object-fit: cover;
  }
`;

type Props = {
  ...ImageType,
};

const HeaderImage = ({ imageUrl, name, url }: Props) => {
  const imgUrl = `${cloudinary}/${String(imageUrl)}`;

  return (
    <ImageWrap>
      <img
        src={imgUrl}
        alt={name != null ? `Taken by ${String(name)}` : undefined}
        data-source-url={url != null ? unsplash(String(url)) : undefined}
      />
    </ImageWrap>
  );
};

export default HeaderImage;
