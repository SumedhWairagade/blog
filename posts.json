{
  "posts": [
    {
      "link": "sinatra-project",
      "title": "Sinatra Project",
      "date": "April 25, 2017",
      "image": "farzad-nazifi-71686.jpg"
    },
    {
      "link": "time-to-hex",
      "title": "How To Convert The Current Time To A Hexadecimal",
      "date": "January 10, 2017",
      "image": "team-ui8-199275.jpg"
    },
    {
      "link": "similarities-between-life-and-code",
      "title": "Similarities Between Life And Code",
      "date": "December 25, 2016",
      "image": "markus-spiske-109588.jpg"
    },
    {
      "link": "rack-key-concepts",
      "title": "Rack Key Concepts",
      "date": "December 7, 2016",
      "image": "michail-sapiton-234263.jpg"
    },
    {
      "link": "javascript-classes-key-concepts",
      "title": "Javascript Classes Key Concepts",
      "date": "November 21, 2016",
      "image": "olu-eletu-134760.jpg"
    },
    {
      "link": "object-lifecycle-cheatsheet",
      "title": "Object Lifecycle Cheatsheet",
      "date": "November 4, 2016",
      "image": "farzad-nazifi-71686.jpg"
    },
    {
      "link": "html5-progress-element",
      "title": "The HTML5 Progress Element is great for blogs",
      "date": "October 14, 2016",
      "image": "wesson-wang-110739.jpg"
    },
    {
      "link": "why-im-learning-software-development",
      "title": "Why I'm Learning Software Development",
      "date": "October 3, 2016",
      "image": "anthony-delanoix-14635.jpg"
    }
  ]
}
