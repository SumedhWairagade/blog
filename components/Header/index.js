import React from 'react';
import PropTypes from 'prop-types';
import { defineMessages, FormattedMessage } from 'react-intl';
import styled from 'styled-components';
import findPost from '../../utils/findPost';
import Button from '../Button';
import Image from './Image';
import Date from './Date';

const messages = defineMessages({
  readMore: {
    id: 'Header.readMore',
    defaultMessage: 'Read More',
  },
});

const Header = styled.header`
  height: 50vh;
  min-height: 50rem;
  max-height: 80rem;
  display: flex;
  justify-content: center;
  align-items: center;
  color: white;
  position: relative;

  @media (orientation: landscape) and (max-height: 500px) {
    max-height: 100vh;
    min-height: 100vh;
    height: 100vh;
  }
`;

const HeaderContent = styled.div`
  z-index: 1;
  text-align: center;
  max-width: 80vw;
`;

export const Title = styled.h1`
  font-size: 5rem;
  margin-bottom: 2rem;
`;

const HeaderWrap = ({ image, title, link, id }) => {
  const post = findPost(id);
  return (
    <Header>
      <HeaderContent>
        <Title>{title || post.title}</Title>
        {!link && post && post.date && <Date date={post.date} />}
        {link && (
          <Button
            text={<FormattedMessage {...messages.readMore} />}
            link={id}
          />
        )}
      </HeaderContent>
      <Image image={image || post.image} />
    </Header>
  );
};

HeaderWrap.defaultProps = {
  link: null,
  id: null,
  title: null,
  image: null,
};

HeaderWrap.propTypes = {
  title: PropTypes.string,
  link: PropTypes.string,
  id: PropTypes.string,
  image: PropTypes.shape({
    imageUrl: PropTypes.string,
    url: PropTypes.string,
    name: PropTypes.string,
  }),
};

export default HeaderWrap;
