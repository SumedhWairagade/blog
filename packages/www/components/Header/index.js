// @flow
import React from 'react';
import styled from 'styled-components';
import dynamic from 'next/dynamic';
import Image from './Image';
import type { ImageType } from '../../types/Post';

const Button = dynamic({
  loader: () => import('../Button'),
  loading: () => null,
});

const HeaderStyles = styled.header.attrs({ 'data-testid': 'header' })`
  height: 50vh;
  min-height: 50rem;
  max-height: 80rem;
  display: flex;
  justify-content: center;
  align-items: center;
  color: white;
  position: relative;

  @media (orientation: landscape) and (max-height: 500px) {
    max-height: 100vh;
    min-height: 100vh;
    height: 100vh;
  }

  .header__content {
    z-index: 1;
    text-align: center;
    max-width: 80vw;
  }

  h1 {
    margin-bottom: 2rem;
    font-size: 3rem;

    @media (min-width: 400px) {
      font-size: 4rem;
    }
  }

  time {
    font-size: 3rem;
    display: block;
  }
`;

type Props = {
  title: string,
  link?: string,
  date?: number,
  image?: ImageType,
  noDate: boolean,
};

const formatDate = new Intl.DateTimeFormat();

const postedTime = new Intl.DateTimeFormat('default', {
  month: 'long',
  day: 'numeric',
  year: 'numeric',
});

const Header = ({ image, link, date, title, noDate }: Props) => {
  const showButton = link != null;

  return (
    <HeaderStyles>
      <div className="header__content">
        <h1>{title}</h1>
        {date != null && noDate === false && (
          <time dateTime={formatDate.format(date)}>
            {postedTime.format(new Date(date))}
          </time>
        )}
        {showButton && <Button text="Read More" link={link} />}
      </div>
      <Image {...image} />
    </HeaderStyles>
  );
};

Header.defaultProps = {
  link: null,
  date: null,
  image: {
    imageUrl: 'paul-volkmer-522844-unsplash.jpg',
    name: 'Paul Volkmer',
    url: 'https://unsplash.com/photos/updW-QUccFE',
  },
};

export default Header;
