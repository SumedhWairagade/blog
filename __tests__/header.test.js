/* eslint-env jest */
import { shallow } from 'enzyme';
import React from 'react';

import Header, { Title } from '../components/Header';
import posts from '../posts.json';

const [latest] = posts;

describe('Header Component', () => {
  it('Header shows latest post', () => {
    const header = shallow(<Header id={latest.id} link={latest.id} />);
    expect(
      header
        .find(Title)
        .render()
        .text()
    ).toBe(latest.title);
  });

  it('Header shows custom text and image', () => {
    const header = shallow(
      <Header
        title="BLOG"
        image={{
          imageUrl:
            'http://localhost:3000/static/images/posts/1*Wmv8hfi_bTHuHyV5CawnCw.jpg',
        }}
      />
    );
    expect(
      header
        .find(Title)
        .render()
        .text()
    ).toBe('BLOG');
  });
});

// describe('With Snapshot Testing', () => {
//   it('App shows "Hello world!"', () => {
//     const component = renderer.create(<App />);
//     const tree = component.toJSON();
//     expect(tree).toMatchSnapshot();
//   });
// });
