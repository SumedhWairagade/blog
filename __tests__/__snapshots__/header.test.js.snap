// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component with Snapshot Testing shows custom text and image 1`] = `
<header
  className="Header-s1qosm5-0 dbhoxp"
>
  <div
    className="Header__HeaderContent-s1qosm5-1 vBaYh"
  >
    <h1
      className="Header__Title-s1qosm5-2 dnbrvZ"
    >
      BLOG
    </h1>
  </div>
  <picture
    className="Image__Picture-s1ifvl2x-0 elFaZh"
  >
    <source
      srcSet="/static/images/posts/http://localhost:3000/static/images/posts/1*Wmv8hfi_bTHuHyV5CawnCw.webp"
      type="image/webp"
    />
    <source
      srcSet="/static/images/posts/http://localhost:3000/static/images/posts/1*Wmv8hfi_bTHuHyV5CawnCw.jpg"
      type="image/jpg"
    />
    <img
      alt=""
      className="Image-s1ifvl2x-1 jXqHhq"
      data-source-url=""
      src="/static/images/posts/http://localhost:3000/static/images/posts/1*Wmv8hfi_bTHuHyV5CawnCw.jpg"
    />
  </picture>
</header>
`;
